src/clusterfuzz/_internal/cron/schedule_fuzz.py

```diff
@@ -0,0 +1,227 @@
+
+import collections
+import random
+import time
+from typing import Dict
+
+from googleapiclient import discovery
+
+from clusterfuzz._internal.base import concurrency
+from clusterfuzz._internal.base import tasks
+from clusterfuzz._internal.base import utils
+from clusterfuzz._internal.config import local_config
+from clusterfuzz._internal.datastore import data_types
+from clusterfuzz._internal.datastore import ndb_utils
+from clusterfuzz._internal.google_cloud_utils import credentials
+from clusterfuzz._internal.metrics import logs
+
+
+def _get_quotas(project, region):
+  gcp_credentials = credentials.get_default()[0]
+  compute = discovery.build('compute', 'v1', credentials=gcp_credentials)
+  return compute.regions().get(  # pylint: disable=no-member
+      region=region, project=project).execute()['quotas']
+
+
+def get_available_cpus(project: str, region: str) -> int:
+  """Gets the number of available CPUs in the current GCE region."""
+  quotas = _get_quotas(project, region)
+
+  # If preemptible quota is not defined, we need to use CPU quota instead.
+  cpu_quota = None
+  preemptible_quota = None
+
+  # Get preemptible_quota and cpu_quota from the list of quotas.
+  for quota in quotas:
+    if preemptible_quota and cpu_quota:
+      break
+    if quota['metric'] == 'CPUS':
+      cpu_quota = quota
+      continue
+    if quota['metric'] == 'PREEMPTIBLE_CPUS':
+      preemptible_quota = quota
+      continue
+  assert preemptible_quota or cpu_quota
+
+  if not preemptible_quota['limit']:
+    quota = cpu_quota
+  else:
+    quota = preemptible_quota
+  assert quota['limit'], quota
+
+  return quota['limit'] - quota['usage']
+
+
+class BaseFuzzTaskScheduler:
+  """Base fuzz task scheduler for any deployment of ClusterFuzz."""
+
+  def __init__(self, num_cpus):
+    self.num_cpus = num_cpus
+
+  def get_fuzz_tasks(self):
+    raise NotImplementedError('Child class must implement.')
+
+  def _get_cpus_per_fuzz_job(self, job_name):
+    del job_name
+    # TODO(metzman): Actually implement this.
+    return 2
+
+
+class FuzzerJob:
+  """Data class that holds more info about FuzzerJobs than the ndb.Models do.
+  Something like this would probably not be needed if we were using SQL and
+  could use joins."""
+
+  def __init__(self, job, project, queue, fuzzer=None, weight=None):
+    self.job = job
+    self.project = project
+    self.queue = queue
+    self.fuzzer = fuzzer
+    self.weight = weight
+
+  def copy(self):
+    return FuzzerJob(
+        job=self.job,
+        project=self.project,
+        queue=self.queue,
+        fuzzer=self.fuzzer,
+        weight=self.weight)
+
+
+class OssfuzzFuzzTaskScheduler(BaseFuzzTaskScheduler):
+  """Fuzz task scheduler for OSS-Fuzz."""
+
+  def get_fuzz_tasks(self) -> Dict[str, tasks.Task]:
+    # TODO(metzman): Handle high end.
+    # A job's weight is determined by its own weight and the weight of the
+    # project is a part of. First get project weights.
+    logs.info('Getting projects.')
```

Comment:
Is it more useful to instead log the total weight we computed?

i.e. "Computed total project CPU weight of X from Y projects"

Reply:
These logs are sort of only for tracking control flow, they probably can be removed now that I know this code works pretty well. Should I?

I don't think the weight thing is important to log because i don't think it will be a meaningful number. The total is just a denominator basically right?